File: quickrtc-flutter-client/lib/services/stream_service.dart
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:logger/logger.dart';
import 'package:quickrtc_flutter_client/models/local_stream_info.dart';
import 'package:quickrtc_flutter_client/models/remote_participant.dart';

/// Options for producing media
class ProduceMediaOptions {
  final MediaStreamTrack? audioTrack;
  final MediaStreamTrack? videoTrack;

  ProduceMediaOptions({
    this.audioTrack,
    this.videoTrack,
  });
}

/// Result of producing media
class ProduceMediaResult {
  final String? audioStreamId;
  final String? videoStreamId;

  ProduceMediaResult({
    this.audioStreamId,
    this.videoStreamId,
  });
}

/// Service for managing media streams
class StreamService {
  static final StreamService _instance = StreamService._internal();
  factory StreamService() => _instance;
  StreamService._internal();

  final Logger _logger = Logger();

  /// Produce media from tracks
  Future<ProduceMediaResult> produceMedia({
    required RTCPeerConnection sendTransport,
    required ProduceMediaOptions options,
    required Function(String kind, Map<String, dynamic> rtpParameters)
        onProduce,
  }) async {
    _logger.d('Producing media');

    String? audioStreamId;
    String? videoStreamId;

    try {
      // Create a local media stream to hold the tracks
      final localStream = await createLocalMediaStream('local_stream');

      // Produce audio if provided
      if (options.audioTrack != null) {
        _logger.d('Adding audio track to transport');
        await localStream.addTrack(options.audioTrack!);
        final audioSender = await sendTransport.addTrack(
          options.audioTrack!,
          localStream,
        );
        final audioParams = await _getRtpParameters(audioSender);
        audioStreamId = await onProduce('audio', audioParams);
        _logger.i('Audio produced with ID: $audioStreamId');
      }

      // Produce video if provided
      if (options.videoTrack != null) {
        _logger.d('Adding video track to transport');
        await localStream.addTrack(options.videoTrack!);
        final videoSender = await sendTransport.addTrack(
          options.videoTrack!,
          localStream,
        );
        final videoParams = await _getRtpParameters(videoSender);
        videoStreamId = await onProduce('video', videoParams);
        _logger.i('Video produced with ID: $videoStreamId');
      }

      return ProduceMediaResult(
        audioStreamId: audioStreamId,
        videoStreamId: videoStreamId,
      );
    } catch (e) {
      _logger.e('Failed to produce media: $e');
      rethrow;
    }
  }

  /// Create local stream info
  Future<LocalStreamInfo> createLocalStreamInfo({
    required String streamId,
    required LocalStreamType type,
    required MediaStreamTrack track,
    required dynamic producer,
  }) async {
    // Create a new MediaStream using helper
    final stream = await createLocalMediaStream(streamId);
    await stream.addTrack(track);

    return LocalStreamInfo(
      id: streamId,
      type: type,
      track: track,
      stream: stream,
      enabled: track.enabled,
      producer: producer,
    );
  }

  /// Stop local stream
  Future<void> stopLocalStream(LocalStreamInfo streamInfo) async {
    _logger.d('Stopping local stream: ${streamInfo.id}');

    try {
      // Stop the track
      await streamInfo.track.stop();

      // Dispose the stream
      streamInfo.stream.dispose();

      _logger.i('Local stream stopped: ${streamInfo.id}');
    } catch (e) {
      _logger.e('Failed to stop local stream: $e');
      rethrow;
    }
  }

  /// Consume participant media
  Future<RemoteParticipant> consumeParticipant({
    required RTCPeerConnection recvTransport,
    required String participantId,
    required String participantName,
    required List<Map<String, dynamic>> consumerParams,
  }) async {
    _logger.d('Consuming participant: $participantId');

    MediaStream? audioStream;
    MediaStream? videoStream;
    dynamic audioConsumer;
    dynamic videoConsumer;

    try {
      for (final params in consumerParams) {
        final kind = params['kind'] as String;

        _logger.d('Creating consumer for $kind');

        // Add transceiver for receiving
        final transceiverInit = RTCRtpTransceiverInit(
          direction: TransceiverDirection.RecvOnly,
        );

        final transceiver = await recvTransport.addTransceiver(
          kind: kind == 'audio'
              ? RTCRtpMediaType.RTCRtpMediaTypeAudio
              : RTCRtpMediaType.RTCRtpMediaTypeVideo,
          init: transceiverInit,
        );

        if (kind == 'audio') {
          audioConsumer = transceiver.receiver;
          final track = transceiver.receiver.track;
          if (track != null) {
            audioStream =
                await createLocalMediaStream('${participantId}_audio');
            await audioStream.addTrack(track);
          }
        } else if (kind == 'video') {
          videoConsumer = transceiver.receiver;
          final track = transceiver.receiver.track;
          if (track != null) {
            videoStream =
                await createLocalMediaStream('${participantId}_video');
            await videoStream.addTrack(track);
          }
        }
      }
      _logger.i('Participant consumed: $participantId');

      return RemoteParticipant(
        participantId: participantId,
        participantName: participantName,
        audioStream: audioStream,
        videoStream: videoStream,
        audioConsumer: audioConsumer,
        videoConsumer: videoConsumer,
        isAudioEnabled: audioStream != null,
        isVideoEnabled: videoStream != null,
      );
    } catch (e) {
      _logger.e('Failed to consume participant: $e');
      rethrow;
    }
  }

  /// Stop consuming participant
  Future<void> stopConsumingParticipant(RemoteParticipant participant) async {
    _logger
        .d('Stopping consumption of participant: ${participant.participantId}');

    try {
      // Dispose audio stream
      if (participant.audioStream != null) {
        participant.audioStream!.dispose();
      }

      // Dispose video stream
      if (participant.videoStream != null) {
        participant.videoStream!.dispose();
      }

      _logger.i('Stopped consuming participant: ${participant.participantId}');
    } catch (e) {
      _logger.e('Failed to stop consuming participant: $e');
      rethrow;
    }
  }

  /// Close consumer
  void closeConsumer(dynamic consumer) {
    _logger.d('Closing consumer');
    // Consumer cleanup if needed
  }

  /// Get RTP parameters from sender
  Future<Map<String, dynamic>> _getRtpParameters(RTCRtpSender sender) async {
    // In flutter_webrtc, we need to construct RTP parameters manually
    // The actual parameters will be negotiated via SDP during peer connection
    final track = sender.track;
    if (track == null) {
      throw Exception('Sender track is null');
    }

    // Return minimal RTP parameters structure
    // The actual negotiation happens through the peer connection's SDP exchange
    return {
      'codecs': [],
      'headerExtensions': [],
      'encodings': [
        {
          'active': true,
          'ssrc': DateTime.now().millisecondsSinceEpoch, // Temporary SSRC
        }
      ],
      'rtcp': {
        'cname': 'flutter-${track.id}',
      },
    };
  }

  /// Reset stream service
  Future<void> reset() async {
    _logger.d('Resetting stream service');
    // Any cleanup needed
    _logger.i('Stream service reset completed');
  }
}

File: quickrtc-react-client/src/api/streamService.ts
import type {
  Device,
  Transport,
  Producer,
  Consumer,
} from "mediasoup-client/lib/types";
import type {
  LocalStreamInfo,
  LocalStreamType,
  ProduceMediaOptions,
  ProduceMediaResult,
  RemoteParticipant,
  RemoteStreamInfo,
} from "../types";
import { socketService } from "./socketService";

/**
 * Enhanced logger for stream service
 */
const logger = {
  info: (method: string, message: string, data?: any) => {
    const timestamp = new Date().toISOString();
    const prefix = `[StreamService:${method}]`;
    if (data) {
      console.info(`${timestamp} ${prefix} ${message}`, data);
    } else {
      console.info(`${timestamp} ${prefix} ${message}`);
    }
  },
  error: (method: string, message: string, error?: any) => {
    const timestamp = new Date().toISOString();
    const prefix = `[StreamService:${method}]`;
    if (error) {
      console.error(`${timestamp} ${prefix} ${message}`, error);
    } else {
      console.error(`${timestamp} ${prefix} ${message}`);
    }
  },
  debug: (method: string, message: string, data?: any) => {
    const timestamp = new Date().toISOString();
    const prefix = `[StreamService:${method}]`;
    if (data) {
      console.debug(`${timestamp} ${prefix} ${message}`, data);
    } else {
      console.debug(`${timestamp} ${prefix} ${message}`);
    }
  },
};

/**
 * Service for managing media streams (local and remote)
 * Handles producers and consumers
 */
export class StreamService {
  /**
   * Produce media from tracks
   */
  public async produceMedia(
    sendTransport: Transport,
    options: ProduceMediaOptions
  ): Promise<ProduceMediaResult> {
    const result: ProduceMediaResult = {};

    try {
      // Produce audio if provided
      if (options.audioTrack) {
        await sendTransport.produce({
          track: options.audioTrack,
          codecOptions: {
            opusStereo: true,
            opusDtx: true,
          },
        });

        result.audioStreamId = `audio-${Date.now()}`;
      }

      // Produce video if provided
      if (options.videoTrack) {
        await sendTransport.produce({
          track: options.videoTrack,
          codecOptions: {
            videoGoogleStartBitrate: 1000,
          },
        });

        const type = options.type || "video";
        result.videoStreamId = `${type}-${Date.now()}`;
      }

      return result;
    } catch (error) {
      console.error("Error producing media:", error);
      throw error;
    }
  }

  /**
   * Create local stream info from track and producer
   */
  public createLocalStreamInfo(
    streamId: string,
    type: LocalStreamType,
    track: MediaStreamTrack,
    producer: Producer
  ): LocalStreamInfo {
    const stream = new MediaStream([track]);
    
    logger.info('createLocalStreamInfo', `Creating local stream info`, {
      streamId,
      type,
      producerId: producer.id,
      trackId: track.id,
      trackKind: track.kind,
    });

    return {
      id: streamId,
      type,
      track,
      producer,
      stream,
      enabled: true,
    };
  }

  /**
   * Stop local stream
   */
  public async stopLocalStream(streamInfo: LocalStreamInfo): Promise<void> {
    const METHOD = 'stopLocalStream';
    try {
      logger.info(METHOD, `Stopping local stream`, {
        streamId: streamInfo.id,
        type: streamInfo.type,
        producerId: streamInfo.producer.id,
        trackId: streamInfo.track.id,
        trackReadyState: streamInfo.track.readyState,
        producerClosed: streamInfo.producer.closed,
      });

      // Stop the track
      logger.debug(METHOD, `Stopping track ${streamInfo.track.id}`);
      streamInfo.track.stop();

      // Close the producer on server
      logger.debug(METHOD, `Closing producer ${streamInfo.producer.id} on server`);
      await socketService.closeProducer(streamInfo.producer.id);

      // Close the producer locally
      logger.debug(METHOD, `Closing producer ${streamInfo.producer.id} locally`);
      streamInfo.producer.close();
      
      logger.info(METHOD, `Successfully stopped ${streamInfo.type} stream`, {
        streamId: streamInfo.id,
        producerId: streamInfo.producer.id,
      });
    } catch (error) {
      logger.error(METHOD, `Error stopping local stream ${streamInfo.id}`, error);
      throw error;
    }
  }

  /**
   * Consume participant media
   */
  public async consumeParticipant(
    device: Device,
    recvTransport: Transport,
    participantId: string,
    participantName: string
  ): Promise<Partial<RemoteParticipant>> {
    try {
      // Get RTP capabilities
      const rtpCapabilities = device.rtpCapabilities;
      if (!rtpCapabilities) {
        throw new Error("Device RTP capabilities not available");
      }

      // Request consumer parameters from server
      const consumerParams = await socketService.consumeParticipantMedia({
        participantId,
        rtpCapabilities,
      });

      if (!consumerParams || consumerParams.length === 0) {
        console.log(`No media to consume from participant ${participantId}`);
        return {};
      }

      const result: Partial<RemoteParticipant> = {
        participantId,
        participantName,
        isAudioEnabled: false,
        isVideoEnabled: false,
        isScreenShareEnabled: false,
        streams: [],
      };

      // Create consumers for each media type
      for (const params of consumerParams) {
        const consumer = await recvTransport.consume({
          id: params.id,
          producerId: params.producerId,
          kind: params.kind,
          rtpParameters: params.rtpParameters,
        });

        // Resume consumer
        await socketService.resumeConsumer(consumer.id);

        // Create media stream
        const stream = new MediaStream([consumer.track]);

        // Determine stream type from server response or fallback to kind
        const serverStreamType = (params as any).streamType;
        let streamType: "audio" | "video" | "screenshare";
        
        if (serverStreamType === 'screenshare') {
          streamType = 'screenshare';
          result.isScreenShareEnabled = true;
        } else if (params.kind === "audio" || serverStreamType === 'audio') {
          streamType = "audio";
          // Keep backward compatibility
          result.audioConsumer = consumer;
          result.audioStream = stream;
          result.isAudioEnabled = true;
        } else {
          // Video camera
          streamType = "video";
          // Keep backward compatibility
          result.videoConsumer = consumer;
          result.videoStream = stream;
          result.isVideoEnabled = true;
        }

        // Add to streams array
        const streamInfo: RemoteStreamInfo = {
          id: `${participantId}-${streamType}-${Date.now()}`,
          type: streamType,
          stream,
          consumer,
          producerId: params.producerId,
        };
        result.streams!.push(streamInfo);
      }

      return result;
    } catch (error) {
      console.error(`Error consuming participant ${participantId}:`, error);
      throw error;
    }
  }

  /**
   * Stop consuming participant media
   */
  public async stopConsumingParticipant(
    participant: RemoteParticipant
  ): Promise<void> {
    try {
      // Close all streams
      if (participant.streams && participant.streams.length > 0) {
        for (const streamInfo of participant.streams) {
          try {
            await socketService.closeConsumer(streamInfo.consumer.id);
            streamInfo.consumer.close();
          } catch (error) {
            console.error(`Error closing consumer ${streamInfo.consumer.id}:`, error);
          }
        }
      } else {
        // Backward compatibility: close legacy consumers
        if (participant.audioConsumer) {
          await socketService.closeConsumer(participant.audioConsumer.id);
          participant.audioConsumer.close();
        }

        if (participant.videoConsumer) {
          await socketService.closeConsumer(participant.videoConsumer.id);
          participant.videoConsumer.close();
        }
      }
    } catch (error) {
      console.error("Error stopping participant consumption:", error);
      throw error;
    }
  }

  /**
   * Close specific consumer (audio or video)
   */
  public closeConsumer(consumer: Consumer): void {
    try {
      consumer.close();
    } catch (error) {
      console.error("Error closing consumer:", error);
    }
  }
}

// Export singleton instance
export const streamService = new StreamService();
